workflow-service/src/main/java/com/redhat/parodos/workflow/WorkFlowDelegate.java

```diff
@@ -41,34 +46,53 @@ public WorkFlowDelegate(BeanWorkFlowRegistryImpl beanWorkFlowRegistry) {
 		this.beanWorkFlowRegistry = beanWorkFlowRegistry;
 	}
 
-	public WorkContext initWorkFlowContext(WorkFlowRequestDTO workFlowRequestDTO) {
+	public WorkContext initWorkFlowContext(WorkFlowRequestDTO workFlowRequestDTO,
+			WorkFlowDefinitionResponseDTO masterWorkFlowDefinitionDto) {
 		WorkContext workContext = new WorkContext();
 
 		if (workFlowRequestDTO.getArguments() != null && !workFlowRequestDTO.getArguments().isEmpty()) {
 			WorkContextDelegate.write(workContext, WorkContextDelegate.ProcessType.WORKFLOW_EXECUTION,
 					workFlowRequestDTO.getWorkFlowName(), WorkContextDelegate.Resource.ARGUMENTS,
 					WorkFlowDTOUtil.convertArgumentListToMap(workFlowRequestDTO.getArguments()));
 		}
-		if (workFlowRequestDTO.getWorks() != null)
-			workFlowRequestDTO.getWorks()
-					.forEach(work -> initWorkContext(workContext, work, workFlowRequestDTO.getWorkFlowName()));
+		if (masterWorkFlowDefinitionDto.getWorks() != null && !masterWorkFlowDefinitionDto.getWorks().isEmpty())
+			masterWorkFlowDefinitionDto.getWorks()
+					.forEach(work -> initWorkContext(workContext,
+							findWorkFromRequest(workFlowRequestDTO, work.getName()), work,
+							masterWorkFlowDefinitionDto.getName()));
 		return workContext;
 	}
 
 	private void initWorkContext(WorkContext workContext, WorkFlowRequestDTO.WorkRequestDTO workRequestDTO,
-			String parentWorkflowName) {
+			WorkDefinitionResponseDTO workDefinitionResponseDTO, String parentWorkflowName) {
 		WorkContextDelegate.write(workContext, WorkContextDelegate.ProcessType.WORKFLOW_EXECUTION,
-				workRequestDTO.getWorkName(), WorkContextDelegate.Resource.PARENT_WORKFLOW, parentWorkflowName);
-		if (workRequestDTO.getArguments() != null && !workRequestDTO.getArguments().isEmpty()) {
+				workDefinitionResponseDTO.getName(), WorkContextDelegate.Resource.PARENT_WORKFLOW, parentWorkflowName);
+
+		if (workRequestDTO != null && workRequestDTO.getArguments() != null
+				&& !workRequestDTO.getArguments().isEmpty()) {
 			WorkContextDelegate.write(workContext,
-					WorkType.WORKFLOW.name().equals(workRequestDTO.getType())
+					WorkType.WORKFLOW.name().equalsIgnoreCase(workRequestDTO.getType())
 							? WorkContextDelegate.ProcessType.WORKFLOW_EXECUTION
 							: WorkContextDelegate.ProcessType.WORKFLOW_TASK_EXECUTION,
 					workRequestDTO.getWorkName(), WorkContextDelegate.Resource.ARGUMENTS,
 					WorkFlowDTOUtil.convertArgumentListToMap(workRequestDTO.getArguments()));
 		}
-		if (workRequestDTO.getWorks() != null)
-			workRequestDTO.getWorks().forEach(work -> initWorkContext(workContext, work, workRequestDTO.getWorkName()));
+		if (workDefinitionResponseDTO.getWorks() != null)
+			workDefinitionResponseDTO.getWorks().forEach(work -> initWorkContext(workContext,
+					findSubWorkFromRequest(workRequestDTO, work.getName()), work, workDefinitionResponseDTO.getName()));
+	}
+
+	private WorkFlowRequestDTO.WorkRequestDTO findWorkFromRequest(WorkFlowRequestDTO workFlowRequestDTO, String name) {
```

Comment:
could you move this method to the DTO, so you can use the method in the dto instead of the service.

Reply:
well, all these methods belongs as methods in the DTO